### ISSUE TITLE
renaming built binary during runtime makes subsequent on-demand imports fail (`ImportError: Loader FrozenImporter cannot handle module`)

### ISSUE THREAD

**Issue by SomberNight** [pyinstaller/pyinstaller#6856]
If a user executes a pyinstaller-built-binary, and while it runs, the user then renames the binary, the running binary will keep running but exhibit surprising (to users) errors. In particular, any subsequent import statement for not-yet-imported modules will fail, as the python code cannot be read from the binary because the binary cannot be found due to the rename.

This affects both onefile and onedir executables. Presumably on all platforms, although I have only manually verified and reproduced on Windows.

Downstream issue: https://github.com/spesmilo/electrum/issues/4072
We have received hundreds of reports for this over the years, and it took a long time to figure out that the cause is this user behaviour.

Consider e.g. this report:
```python
Traceback (most recent call last):
  File "PyInstaller\loader\pyimod03_importers.py", line 298, in get_code
  File "PyInstaller\loader\pyimod02_archive.py", line 290, in extract
  File "PyInstaller\loader\pyimod02_archive.py", line 87, in __enter__
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\pferd\\Downloads\\electrum-4.2.2-portable.exe'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "electrum\gui\qt\main_window.py", line 3203, in settings_dialog
  File "PyInstaller\loader\pyimod03_importers.py", line 478, in exec_module
  File "PyInstaller\loader\pyimod03_importers.py", line 300, in get_code
ImportError: Loader FrozenImporter cannot handle module electrum.gui.qt.settings_dialog
```
- this user downloaded `electrum-4.2.2-portable.exe`, a onefile pyinstaller-built binary (built with pyinst 4.10)
- they then started the binary and started interacting with the GUI
- they then renamed or moved `electrum-4.2.2-portable.exe`
- interacting with the GUI again, they tried to open the "settings dialog", which is imported on-demand,
- this raises an `ImportError` -- the pyinstaller loader cannot find the code as it has been moved

Note that they might have the program running for hours, and forget that they have renamed the exe. If a long time has passed between renaming the exe and triggering the ImportError, it is not obvious that the events are linked.

-----

Looking at the long list of reports in [the linked issue](https://github.com/spesmilo/electrum/issues/4072), note that there are also some for macOS, although most are for Windows. I suppose Linux is also affected but we don't use pyinstaller for Linux binaries.

Not sure about a cross-platform solution, but I think a solution for Windows could be to have the binary "open" in read mode at runtime. On Windows, if one process has a file open the OS forbids other processes to rename the file.

**Comment by bwoodsend** [comment]
> but I think a solution for Windows could be to have the binary "open" in read mode at runtime.

I didn't think that it was actually possible to execute an executable in Windows without the binary being open and therefore being locked.

**Comment by rokm** [comment]
Looks like our archive reader is deliberately not leaving an open file handle to the executable, otherwise that alone would prevent the executable (and, by extension, onedir bundle) from being renamed. That said, in addition to frozen importer not being able to read from the embedded archive, renaming the executable while the program is running is also bad news for `multiprocessing` and other users of `sys.executable`, because the latter will continue pointing to the original (and now non-existent) file name.

While keeping an open file handle would prevent users from renaming the executable while it is running, I don't think we should be changing the current behavior. Because the moment we do, we'll get regression reports from people who implemented some sort of auto-update scheme, and suddenly won't be able to overwrite their program files. On the other hand, if you think your program should be protecting users from themselves, you can easily do an `open(sys.executable, 'rb')` at the beginning of your program, and lock the executable from being renamed. But to be perfectly honest, if this is supposed to be a long-running program, it should have an installer that puts the executable somewhere where users cannot tamper with it in the first place...

**Comment by bwoodsend** [comment]
If you left the handle for the exe open then try to move the whole bundle, would Windows abort before moving anything or would it start moving files then stop when it hits the exe? If it's the latter then presumably you're screwed anyway.

This sounds so self inflicted. I don't know of any binary executable that handles having the floor pulled out from under its feet gracefully. We have a custom issue label called `play-stupid-games-win-stupid-rewards`. If I were you, I'd be slapping that label on each of your hundred reports.

If you're users are that determined to do this then catch and raise it to make the error less cryptic:

```python
try:
    # main code here
except ImportError as ex:
    if ex.__cause__ and isinstance(ex.__cause__, FileNotFoundError):
        # Tell you users what they have done wrong and exit gracefully.
    else:
        raise
```
I guess that it wouldn't hurt to put some equivalent logic into PyInstaller's custom import loader.

**Comment by SomberNight** [comment]
> But to be perfectly honest, if this is supposed to be a long-running program, it should have an installer that puts the executable somewhere where users cannot tamper with it in the first place...

Well, the application is not necessarily what you would call a long-running program, but I suppose some people just forget to close it, or have it running in the background sometimes.

In any case, for Windows, we distribute both onefile and onedir (installed with NSIS) binaries. Some people specifically asked for "portable" onefile binaries where even the data directory created by the application is just a standalone folder put next to the binary, so that they can e.g. put both the exe and its data on a USB flash drive.

For macOS, in the past we used to build onefile binaries but at some point switched to onedir binaries due to performance issues on macOS 10.15 where the OS would try to scan and upload via the network to Apple's Gatekeeper data about every file unpacked during app start -- with onedir binaries there is no (or much less) unpacking, making them significantly faster. 
Also, from a user's point of view, on macOS, there is basically no difference between a onefile and a onedir binary, unlike on Windows.

> If you left the handle for the exe open then try to move the whole bundle, would Windows abort before moving anything or would it start moving files then stop when it hits the exe? If it's the latter then presumably you're screwed anyway.

Good point. However, note that for onefile binaries this is not applicable. For onedir binaries, in our case, we install them via NSIS into "Program Files". I expect it is mostly the users who download the onefile binaries who move them around.

In particular, if you look at the example report in my OP, they downloaded a onefile exe, it got put into their `Downloads` folder, they started it from there, then they likely moved it, e.g. to the `Desktop`, and then the `ImportError` is raised. Not so unlikely behaviour I would argue.

> This sounds so self inflicted. I don't know of any binary executable that handles having the floor pulled out from under its feet gracefully. We have a custom issue label called `play-stupid-games-win-stupid-rewards`. If I were you, I'd be slapping that label on each of your hundred reports

We only have a single issue about this. The only reason we know about the problem is because we have implemented a "crash reporter" as part of the program. If there is an uncaught exception in any thread, we catch it and pop a window offering the user to send a crash report to an https endpoint to a bot. The bot then aggregates the reports based on the traceback into github issues, creating new issues for new bugs and editing the OP of existing issues for existing bugs. This particular github issue and exception trace has 258 individual reports atm. The URL for the github issue is also shown to the user when sending the report.
None of these 258 users commented on the issue, and no one manually opened other issues. Based on the behaviour of renaming the running exe, I expect none of these users are technical. If not for the semi-automated crash reporting, we would have never learned about this.

> If you're users are that determined to do this then catch and raise it to make the error less cryptic:
> ```python
> try:
>     # main code here
> except ImportError as ex:
>     if ex.__cause__ and isinstance(ex.__cause__, FileNotFoundError):
>         # Tell you users what they have done wrong and exit gracefully.
>     else:
>         raise
> ```

We would have to do this for every import, ctypes dll load, etc, scattered around the codebase. Also, who knows if our dependencies do on-demand imports themselves in any API call. This is simply not realistic.

> On the other hand, if you think your program should be protecting users from themselves, you can easily do an `open(sys.executable, 'rb')` at the beginning of your program, and lock the executable from being renamed

Right, I will likely do this after a bit of experimenting.
Small note that the return value needs to be saved in a global to avoid the gc, for it to work. E.g. `_file = open(sys.executable, 'rb')`.
To be clear, this only works on Windows, right?

**Comment by rokm** [comment]
> Not so unlikely behaviour I would argue.

Even if not uncommon, this still strikes me as a pure user-behavior problem, and one that should be addressed by a note informing the users that they should not be moving/renaming the portable executable while it is running. 

> Small note that the return value needs to be saved in a global to avoid the gc, for it to work. E.g. `_file = open(sys.executable, 'rb')`.

Yeah, the returned handle needs to be stored somewhere in order to prevent gc from kicking in.

>  To be clear, this only works on Windows, right?

Indeed; keeping the file handle open won't prevent the executable being renamed on other OSes.

**Comment by SomberNight** [comment]
>>> On the other hand, if you think your program should be protecting users from themselves, you can easily do an `open(sys.executable, 'rb')` at the beginning of your program, and lock the executable from being renamed
>>
>> Small note that the return value needs to be saved in a global to avoid the gc, for it to work. E.g. `_file = open(sys.executable, 'rb')`.
>
> Yeah, the returned handle needs to be stored somewhere in order to prevent gc from kicking in.

I've tested a bit more now and can confirm this seems to work well on Windows.

-----

Feel free to close this issue if you deem it "wontfix". At least now it's ~documented and comes up on search engines.

**Comment by bwoodsend** [comment]
> We would have to do this for every import, ctypes dll load, etc, scattered around the codebase. Also, who knows if our dependencies do on-demand imports themselves in any API call. This is simply not realistic.

No you wouldn't. Just stick it at the very top level of your code then exceptions raised during any chain of imports/ctypes loads will propagate up the stack. Injecting the `except:` block [here](https://github.com/spesmilo/electrum/blob/5e0df77df6b06330284e2c15e188fbb4a9865fdb/run_electrum#L419) would do. (Edit: I see there's process forking involved in there so it probably won't do in your case.)

> Feel free to close this issue if you deem it "wontfix". At least now it's ~documented and comes up on search engines.

I think I will catch this in our import loader then raise a more descriptive exception. It's a niche case born of misuse but it'll only be a few lines of code.

**Comment by bwoodsend** [comment]
Addressed by #6884.
